openmc/deplete/pool.py

```diff
@@ -59,6 +64,52 @@ def deplete(func, chain, x, rates, dt, matrix_func=None):
     else:
         matrices = map(matrix_func, repeat(chain), rates, fission_yields)
 
+    if msr is not None:
+        # Calculate removal rate terms as diagonal matrices
+        removals = map(chain.form_rr_term, repeat(msr), msr.burnable_mats)
+        # Subtract removal rate terms from Bateman matrices
+        matrices = [matrix - removal for (matrix, removal) in zip(matrices, removals)]
+
+        if len(msr.index_transfer) > 0:
+            # Calculate transfer rate terms as diagonal matrices
+            transfers = list(map(chain.form_rr_term, repeat(msr),
+                                msr.index_transfer))
+
+            # Combine all matrices together in a single matrix of matrices
+            # to be solved in one-go
+            n_rows = n_cols = len(msr.burnable_mats)
+            rows = []
+            rows_cols = \
+                np.array(np.meshgrid(range(n_rows),
+                                     range(n_cols))).T.reshape(-1, 2)
+            last_col = n_cols - 1
+            cols = []
+            for (row, col) in rows_cols:
+                if row == col:
+                    # Fill the diagonals with the Bateman matrices
+                    cols.append(matrices[row])
+                elif (msr.burnable_mats[row], msr.burnable_mats[col]) in \
+                                msr.index_transfer:
+                    index = list(msr.index_transfer).index( \
+                                (msr.burnable_mats[row], msr.burnable_mats[col]))
+                    # Fill the off-diagonals with the transfer matrices
+                    cols.append(transfers[index])
+                else:
+                    cols.append(None)
+                if col == last_col:
+                    rows.append(cols)
+                    cols = []
+            matrix = bmat(rows)
+
+            # Concatenate vectors of nuclides in one
+            _x = np.concatenate([xx for xx in x])
+            x_result = func(matrix, _x, dt)
```

Comment:
We only use variable names with a leading underscore for attributes of a class that are "private", so I would recommend using something else here (`x_multi`, `x_large`, `x_full`, or whatever you think best describes it)